pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -147,48 +147,62 @@ func BuildSidecarVirtualHostWrapper(routeCache *Cache, node *model.Proxy, push *
 	serviceRegistry map[host.Name]*model.Service,
-	mostSpecificWildcardIndex map[host.Name]types.NamespacedName,
+	mostSpecificWildcardVsIndex map[host.Name]types.NamespacedName,
 ) ([]string, []*model.Service) {
 	// TODO: A further optimization would be to completely rely on the index and not do the loop below
 	// However, that requires assuming that serviceRegistry never got filtered after the
 	// egressListener was created.
 	rule := virtualService.Spec.(*networking.VirtualService)
-	hosts := make([]string, 0)
+	// Stores VS hosts that don't correspond to services in the registry
+	// Currently, the only use for this list is to enable VirtualService configuration to affect
+	// traffic to hosts outside of the service registry (e.g. google.com) on port 80
+	nonServiceHosts := make([]string, 0)
+	// Stores services for this VirtualService that are in the registry (based on hostname)
 	servicesInVirtualService := make([]*model.Service, 0)
 	wchosts := make([]host.Name, 0)
 
 	// As a performance optimization, process non wildcard hosts first, so that they can be
 	// looked up directly in the service registry map.
 	for _, hostname := range rule.Hosts {
 		vshost := host.Name(hostname)
-		if !vshost.IsWildCarded() {
-			if svc, exists := serviceRegistry[vshost]; exists {
-				servicesInVirtualService = append(servicesInVirtualService, svc)
-			} else {
-				hosts = append(hosts, hostname)
-			}
-		} else {
-			// Add it to the wildcard hosts so that they can be processed later.
+		if vshost.IsWildCarded() {
+			// We'll process wild card hosts later
 			wchosts = append(wchosts, vshost)
+			continue
+		}
+		if svc, exists := serviceRegistry[vshost]; exists {
+			servicesInVirtualService = append(servicesInVirtualService, svc)
+		} else {
+			nonServiceHosts = append(nonServiceHosts, hostname)
 		}
 	}
 
 	// Now process wild card hosts as they need to follow the slow path of looping through all Services in the registry.
 	for _, hostname := range wchosts {
 		if model.UseGatewaySemantics(virtualService) {
-			hosts = append(hosts, string(hostname))
+			nonServiceHosts = append(nonServiceHosts, string(hostname))
 			continue
 		}
-		// Say this VS's host is *.global and there's another VS with host *.foo.global
+		// foundSvcMatch's only purpose is to make sure we don't add hosts that correspond to services
+		// to the list of non-serviceregistry hosts
 		foundSvcMatch := false
-		// Say we have Services *.foo.global, *.bar.global
 		for svcHost, svc := range serviceRegistry {
-			vs, ok := mostSpecificWildcardIndex[svcHost]
+			// First, check if this service matches the VS host.
+			// If it does, then we never want to add it to the nonServiceHosts list.
+			// The result is OR'd so we don't overwrite a previously true value
+			foundSvcMatch = foundSvcMatch || svcHost.Matches(hostname)
+			if !foundSvcMatch {
+				// If the wildcard doesn't even match this service, it won't be in the index
+				continue
+			}
```

Comment:
The OR logic looks correct in the global context however when considered in the context of the loop I think this may not be exactly what we want because once a match is found were going to check all subsequent svcHost from the registry against the mostSpecific... logic below even if we should probably skip it. My names below are probably bad but I'll leave a suggestion with one way I think addresses it below.
```suggestion
                        // localMatch is tracking the match found within this iteration of the loop
                        localMatch := svcHost.Matches(hostname)
                        // foundSvcMatch is tracking in the wider context whether or not ANY match was found during an iteration
			foundSvcMatch = foundSvcMatch || localMatch
			if !localMatch {
				// If the wildcard doesn't even match this service, it won't be in the index
				continue
			}
```

Reply:
Nice find! +1 to your summary, I like this suggestion